cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java

```diff
@@ -0,0 +1,755 @@
+    runsUrl = getRunsUrl(TEST_NAMESPACE2, PAUSE_RESUME_WORKFLOW_APP, PAUSE_RESUME_WORKFLOW, "suspended");
+    scheduleHistoryRuns(10, runsUrl, 0);
+
+    // Allow currently executing actions to complete
+    forkedSimpleActionDoneFile.createNewFile();
+    anotherForkedSimpleActionDoneFile.createNewFile();
+
+    // Workflow should have zero actions running
+    currentRunningProgramsExpected = 0;
+    checkCurrentRuns(10, versionedUrl, currentRunningProgramsExpected);
+
+    runsUrl = getRunsUrl(TEST_NAMESPACE2, PAUSE_RESUME_WORKFLOW_APP, PAUSE_RESUME_WORKFLOW, "suspended");
+    scheduleHistoryRuns(10, runsUrl, 0);
+
+    Assert.assertTrue(!lastSimpleActionFile.exists());
+
+    resumeWorkflow(programId, runId, 200);
+
+    Assert.assertEquals("RUNNING",
+                        getRunnableStatus(TEST_NAMESPACE2, PAUSE_RESUME_WORKFLOW_APP,
+                                          ProgramType.WORKFLOW.getCategoryName(), PAUSE_RESUME_WORKFLOW));
+
+    while (!lastSimpleActionFile.exists()) {
+      TimeUnit.SECONDS.sleep(1);
+    }
+
+    currentRunningProgramsExpected = 1;
+    checkCurrentRuns(10, versionedUrl, currentRunningProgramsExpected);
+
+    lastSimpleActionDoneFile.createNewFile();
+
+    runsUrl = getRunsUrl(TEST_NAMESPACE2, PAUSE_RESUME_WORKFLOW_APP, PAUSE_RESUME_WORKFLOW, "completed");
+    scheduleHistoryRuns(10, runsUrl, 0);
+
+    Assert.assertEquals("STOPPED",
+                        getRunnableStatus(TEST_NAMESPACE2, PAUSE_RESUME_WORKFLOW_APP,
+                                          ProgramType.WORKFLOW.getCategoryName(), PAUSE_RESUME_WORKFLOW));
+
+  }
+
+  @Category(XSlowTests.class)
+  @Test
+  public void testMultipleWorkflowInstances() throws Exception {
+    // create app in default namespace so that v2 and v3 api can be tested in the same test
+    String defaultNamespace = "default";
+    HttpResponse response = deploy(ConcurrentWorkflowApp.class, Constants.Gateway.API_VERSION_3_TOKEN,
+                                   defaultNamespace);
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+
+    Assert.assertEquals(200, resumeSchedule(defaultNamespace, APP_WITH_CONCURRENT_WORKFLOW,
+                                            APP_WITH_CONCURRENT_WORKFLOW_SCHEDULE_1));
+    Assert.assertEquals(200, resumeSchedule(defaultNamespace, APP_WITH_CONCURRENT_WORKFLOW,
+                                            APP_WITH_CONCURRENT_WORKFLOW_SCHEDULE_2));
+
+    Map<String, String> propMap = Maps.newHashMap();
+    propMap.put(ProgramOptionConstants.CONCURRENT_RUNS_ENABLED, "true");
+    PreferencesStore store = getInjector().getInstance(PreferencesStore.class);
+    store.setProperties(defaultNamespace, APP_WITH_CONCURRENT_WORKFLOW, ProgramType.WORKFLOW.getCategoryName(),
+                        CONCURRENT_WORKFLOW_NAME, propMap);
+
+    String runsUrl = getRunsUrl(defaultNamespace, APP_WITH_CONCURRENT_WORKFLOW, CONCURRENT_WORKFLOW_NAME, "running");
```

Comment:
Don't generate URL and call `scheduleHistoryRuns`. Have a `scheduleHistoryRun(Id.Program programId)` to do the job (URL generation should be in that method). I've been fixing that in the AppFabricTestBase (go to that class and see how it should be. E.g. Look at `startProgram` or `waitState` methods there). We should stop doing things like this in new code.